crates/solver/src/solver/naive_solver/multi_order_solver.rs

```diff
@@ -316,16 +316,37 @@ fn is_valid_solution(solution: &Settlement) -> bool {
         data,
         sell_token_price,
         buy_token_price,
-    } in solution.encoder.user_trades()
+    } in solution.encoder.all_trades()
     {
         let order = &data.order.data;
+
+        // Check execution respects individual order's limit price
         match (
             order.sell_amount.checked_mul(sell_token_price),
             order.buy_amount.checked_mul(buy_token_price),
         ) {
             (Some(sell_volume), Some(buy_volume)) if sell_volume >= buy_volume => (),
             _ => return false,
         }
+
+        // Check individual order's execution price satisfies uniform clearing price
+        // E.g. liquidity orders may have a different executed price.
+        match (
+            solution
+                .encoder
+                .clearing_prices()
+                .get(&order.buy_token)
+                .map(|clearing_sell_price| clearing_sell_price.checked_mul(sell_token_price)),
+            solution
+                .encoder
+                .clearing_prices()
```

Comment:
nit: a little less verbose
```suggestion
        let clearing_prices = solution.encoder.clearing_prices();
        match (
            clearing_prices
                .get(&order.buy_token)
                .map(|clearing_sell_price| clearing_sell_price.checked_mul(sell_token_price)),
            clearing_prices
```